lib/Dialect/Arc/Transforms/SplitLoops.cpp

```diff
@@ -0,0 +1,358 @@
+  for (auto operand : block.getTerminator()->getOperands()) {
+    if (auto blockArg = dyn_cast<BlockArgument>(operand)) {
+      outputs.push_back({true, blockArg.getArgNumber(), 0});
+      continue;
+    }
+    auto &operandSplit = getSplit(coloring.lookup(operand.getDefiningOp()));
+    auto resultIdx = operandSplit.exportValue(operand);
+    outputs.push_back({false, resultIdx, operandSplit.index});
+  }
+
+  // Create the final `arc.output` op for each of the splits.
+  for (auto &split : splits)
+    split->builder.create<arc::OutputOp>(loc, split->exportedValues);
+}
+
+/// Get or create the split for a given operation color.
+Split &Splitter::getSplit(const APInt &color) {
+  auto &split = splitsByColor[color];
+  if (!split) {
+    auto index = splits.size();
+    LLVM_DEBUG(llvm::dbgs()
+               << "- Creating split " << index << " for " << color << "\n");
+    split = std::make_unique<Split>(context, index, color);
+    splits.push_back(split.get());
+  }
+  return *split;
+}
+
+//===----------------------------------------------------------------------===//
+// Pass Implementation
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct SplitLoopsPass : public SplitLoopsBase<SplitLoopsPass> {
+  void runOnOperation() override;
+  void splitArc(Namespace &arcNamespace, DefineOp defOp,
+                ArrayRef<StateOp> arcUses);
+  void replaceArcUse(StateOp arcUse, ArrayRef<DefineOp> splitDefs,
+                     ArrayRef<Split *> splits, ArrayRef<ImportedValue> outputs);
+  LogicalResult ensureNoLoops(ArrayRef<StateOp> allArcUses);
+};
+} // namespace
+
+void SplitLoopsPass::runOnOperation() {
+  auto module = getOperation();
+
+  // Collect all arc definitions.
+  Namespace arcNamespace;
+  DenseMap<StringAttr, DefineOp> arcDefs;
+  for (auto arcDef : module.getOps<DefineOp>()) {
+    arcNamespace.newName(arcDef.getSymName());
+    arcDefs[arcDef.getSymNameAttr()] = arcDef;
+  }
+
+  // Collect all arc uses and determine which arcs we should split.
+  SetVector<DefineOp> arcsToSplit;
+  DenseMap<DefineOp, SmallVector<StateOp>> arcUses;
+  SetVector<StateOp> allArcUses;
+
+  module.walk([&](StateOp stateOp) {
+    auto sym = stateOp.getArcAttr().getAttr();
+    auto defOp = arcDefs.lookup(sym);
+    arcUses[defOp].push_back(stateOp);
+    allArcUses.insert(stateOp);
+    if (stateOp.getLatency() == 0 && stateOp.getNumResults() > 1)
+      arcsToSplit.insert(defOp);
+  });
+
+  // Split all arcs with more than one result.
+  // TODO: This is ugly and we should only split arcs that are truly involved in
+  // a loop. But detecting the minimal split among the arcs is fairly
+  // non-trivial and needs a dedicated implementation effort.
+  for (auto defOp : arcsToSplit)
+    splitArc(arcNamespace, defOp, arcUses[defOp]);
+
+  // Ensure that there are no loops through arcs remaining.
+  if (failed(ensureNoLoops(allArcUses.getArrayRef())))
```

Comment:
Maybe I'm missing something, but don't we have to update `allArcUses` according to the IR changes in `splitArc`?

Reply:
Oops, excellent point. Fixed that. I also added an error test that sanity checks this loop detection.